### ISSUE TITLE
Restarting master break mounted client

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#601]
Restarting a master (with no shadow takeover) will break any mounted client:

```
Oct  3 15:23:02 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
Oct  3 15:23:02 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
Oct  3 15:23:04 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
Oct  3 15:23:04 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
Oct  3 15:23:06 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
Oct  3 15:23:06 x mfsmount[15661]: master: register error (read header: ESUCCESS (Success))
```

When this happens, it's also impossible to unmount the client with `unmount /mnt`.
I have to kill `mfsmount` process and even after this the mountpoint is still stale, shown as mounted:
```
# mount | grep mfs
mfs#mfsmaster:9421 on /mnt type fuse (rw,nosuid,nodev,noexec,noatime,user_id=0,group_id=0,default_permissions,allow_other)
```

but not working (obviously):
```
# ls -la /mnt
ls: cannot access '/mnt': Transport endpoint is not connected
```

and still can't be remounted:

```
~# mfsmount -o big_writes,nonempty,nosuid,noexec,nodev,noatime,mfswritecachesize=16384,cacheexpirationtime=3600 /mnt
fuse: bad mount point `/mnt': Transport endpoint is not connected
see: mfsmount -h for help
```

**Comment by guestisp** [comment]
In my case, the only way to fix this is rebooting the node or lazy unmounting the mount point (`umount --lazy`) and wait for some unknown timeout before rerunning `mfsmount` (if done immediatly after the unmount, mount doesn't mount anything :)

**Comment by Blackpaw** [comment]
You mentioned earlier that you were using an out of date client - maybe thats the problem? what version mfsmount and mfsmaster are you using?

**Comment by guestisp** [comment]
Currently, i'm running the latest versions everywhere.

**Comment by 4Dolio** [comment]
That sounds normal. Only when a shadow takes over do their shared sessions persist.  You will need to umount -fl and mount again.

**Comment by Blackpaw** [comment]
That surprises me - mfsmount cannot reconnect to the same master server?

**Comment by guestisp** [comment]
No, even after some minutes. I'll try again shortly

**Comment by guestisp** [comment]
No, it doesn't reconnect. Tried now. Any operation on the mount point (in my case `/mnt`) will hang.

Maybe I have to set a timeout somewhere ?

**Comment by guestisp** [comment]
Yes, there is a timeout, first error: `Oct  3 17:24:19` reconnection at `Oct  3 17:26:19`

Exactly 120 seconds, but I don't see any reference in `/etc/lizardfs`:

```
# grep 120 /etc/lizardfs/*
# 
```

**Comment by psarna** [comment]
All timeouts and mount options can be listed via `mfsmount -h`

**Comment by guestisp** [comment]
I don't see any timeout regarting this.

**Comment by 4Dolio** [comment]
That is strange that you can not get the mfsmount client to die...  umount -fl /mountpoint usually does it for me, maybe have to do it a few times until it says "not mounted".  Perhaps you have some object still locked or open underneath your mountpoint?

I would not expect a mount to survive a restart of the server, unless it has robust reconnection.  If you stop an nfs export server you will have to fix all the clients somehow, I believe.  LFS acts much like that if you are not failing over to a shadow which preserves the servers client export table.

**Comment by 4Dolio** [comment]
I do not think timeouts will help here. Your sessions are all dead and can not be recovered by a timeout reconnect.. A new session has to be created by a new mount..  Pretty sure anyway.

**Comment by guestisp** [comment]
> umount -fl /mountpoint usually does it for me

you are using the lazy unmont (`-l`). This is not a solution, lazy unmount should not be used in normal situation.

As wrote above, it's a timeout issue, exactly after 120 seconds, the mount point gets connected again.
How can I tune this settings? I haven't find any related parameters in `mfsmount`

**Comment by 4Dolio** [comment]
but you began by restarting your master and dropping the client session state tables.  So the mount can not recover with a reconnect like it could had a shadow taken over and preserved the sessions.  I guess..

All that comes to mind is:
> MASTER_RECONNECTION_DELAY
> delay in seconds before trying to reconnect to metadata server after disconnection (default is 1)
> MASTER_TIMEOUT
> timeout (in seconds) for metadata server connections (default is 60)

I dont see anything from the client side? Shrugs. But again, i dont think it would matter. You have to do extraordinary things to kill the client which is forever lost without a valid session. Thus --force and --lazy and it should cleanly die, any activity to it sould be expected to cause hung io.
